.ci-cd/Dockerfile.cpu

```diff
@@ -31,8 +33,25 @@ RUN pip3 install  pre-commit==1.17.0 \
     pylint==2.3.1 \
     yapf==0.28.0
 
+# Ubuntu 18.04 officially supports git version 2.17 (through apt-get),
+# but github CI checkout@v2 requires a higher version of git (>=2.18).
+# Here we install git 2.18 from source.
+
+# install some pre-requisites for building git from source
```

Comment:
Can do it more formally from original PPA?

```bash
add-apt-repository ppa:git-core/ppa
apt update
apt install git
```

We probably only need to add the `ppa` if there is `apt install git` somewhere.

Reply:
> Can do it more formally from original PPA?
> 
> ```shell
> add-apt-repository ppa:git-core/ppa
> apt update
> apt install git
> ```
> 
> We probably only need to add the `ppa` if there is `apt install git` somewhere.

This is the first solution I tried, as follows:
```
# for add-apt-repository
RUN apt-get install software-properties-common -y
RUN add-apt-repository ppa:git-core/ppa && apt update -y && apt install git
```
but does not work.

Eventually will get an ``apt_pkg`` related error, which is potentially related to our python3.7 version on ubuntu 18.04.
Build from source takes some time but is cleaner.

Reply:
Ah okay. Sorry I don't have Ubuntu to try. Yep building from source is more consistent and deterministic, so in this case I am supporting building from source.